reconcile/terraform_init/integration.py

```diff
@@ -0,0 +1,165 @@
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.unleash import get_feature_toggle_state
+from reconcile.utils.vcs import VCS
+
+QONTRACT_INTEGRATION = "terraform-init"
+QONTRACT_INTEGRATION_VERSION = make_semver(1, 0, 0)
+
+
+class TerraformInitIntegrationParams(PydanticRunParams):
+    account_name: str | None
+    state_tmpl_resource: str = "/terraform-init/terraform-state.yml"
+    template_collection_root_path: str = "data/templating/collections/terraform-init"
+
+
+class TerraformInitIntegration(
+    QontractReconcileIntegration[TerraformInitIntegrationParams]
+):
+    """Initialize AWS accounts for Terraform usage."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+        return {
+            "accounts": [
+                account.dict() for account in self.get_aws_accounts(query_func)
+            ],
+        }
+
+    def get_aws_accounts(
+        self, query_func: Callable, account_name: str | None = None
+    ) -> list[AWSAccountV1]:
+        """Return all AWS accounts with terraform username but no terraform state set."""
+        return [
+            account
+            for account in aws_accounts_query(query_func).accounts or []
+            if integration_is_enabled(self.name, account)
+            and (not account_name or account.name == account_name)
+            and account.terraform_username
+            and not account.terraform_state
+        ]
+
+    def render_state_collection(
+        self, template: str, bucket_name: str, account: AWSAccountV1
+    ) -> str:
+        return jinja2.Template(
+            template,
+            undefined=jinja2.StrictUndefined,
+            trim_blocks=True,
+            lstrip_blocks=True,
+            keep_trailing_newline=True,
+        ).render({
+            "account_name": account.name,
+            "bucket_name": bucket_name,
+            "region": account.resources_default_region,
+            "timestamp": int(datetime.now(tz=timezone.utc).timestamp()),
+        })
+
+    def reconcile_account(
+        self,
+        account_aws_api: AWSApi,
+        merge_request_manager: MergeRequestManager,
+        dry_run: bool,
+        state_collection: str,
+        bucket_name: str,
+        account: AWSAccountV1,
+    ) -> None:
+        logging.info("Creating bucket '%s' for account '%s'", bucket_name, account.name)
+        if not dry_run:
+            # the creation of the bucket is idempotent
+            account_aws_api.s3.create_bucket(
+                name=bucket_name, region=account.resources_default_region
+            )
```

Comment:
in previous setup, there is a bucket policy attached, is that no needed or we do it in another step?

Reply:
We don't need this extra policy. The new bucket is owned by the creator, which is the future terraform user. In the previous setup, the root user owns the bucket, therefore an extra policy is needed.

Reply:
I think that extra policy is used to limit access to that bucket to terraform user only, like even I use SSO login to that aws account, I can't do anything to that bucket including list objects.

Reply:
Do we need this type of protection? Why should we limit the admin users?

Reply:
I don't know, quote the [doc](https://gitlab.cee.redhat.com/service/app-interface/-/tree/master/docs/aws/terraform)

>The s3 backend is used to share state and is set up with a bucket policy allowing strict access as the state files have highly sensitive content including keys.

Reply:
Ok, I'll leave it as it is. Otherwise, we would need to use the root user (which we may not have access to) to operate on those buckets.